src/events/rematch.js

```diff
@@ -0,0 +1,61 @@
+// Event: leave
+const Games = require('../server/games.service');
+const Users = require('../server/users.service');
+const Rooms = require('../server/rooms.service');
+const Server = require('../server/server.service');
+const Redis = require('../server/redis.service');
+const debug = require('debug')('have_you_heard');
+
+// Initialize event listener
+module.exports = function(socket) {
+    socket.on('rematch', async () => {
+        let userID = `user_${socket.id}`
+
+        await Redis.getIO(async (redisIO) => {
+
+            let user = await Users.get(redisIO, userID);
+
+            // Check if the user exists
+            if (!user) {
+                console.error(`User ${userID} not found`);
+                return;
+            }
+
+            // Check if the user is in a room
+            if (!user.room) {
+                console.error(`User ${userID} is not in a room`);
+                return;
+            }
+
+            // Check if the user is still in a game
+            if (!user.game) {
+                console.error(`User is not in a game`);
+                return;
+            }
+
+            await Games.rematch(redisIO, userID, user.game, async (game) => {
+                if (game) {
+                    // All players reacted to rematch
+                    if (game.waitList.length == 0) {
```

Comment:
I think the `leave` event needs to check if `wailist` length is 0 too. Will test.

Reply:
Yeap, if the last person to answer replies `leave`. The Game is not ended, and the owner of the room cannot start the next game.

Reply:
Right, that is the problem.

But now looking at how we designed the rematch, I think it is bad. The worst part is reusing the `leave` message for a different purpose, which creates ambiguity and the need to track in which stage of the game we are. Is the user simply leaving the room or we are handling a rematch?

I propose we refactor the rematch to use only the `rematch` message, passing a `bool` if the user wants to stay or leave. This way we won't have the ambiguity of what the `leave` means and handle all the rematch-related messages in a single place.

Reply:
OK, I think I fixed in a simpler way:
- When `rematch` is received, remove the player from the game
- When `leave` is received, remove the player from the game and from the room.

While the game exists, the owner cannot start another game. When the game becomes empty it is deleted immediately.

So, no need to change the client side, and far less code and complexity added.

Reply:
Now the owner of the room can start the game before all players have answered, :joy:.

We can make changes on the client to emit `rematch` with a `bool`value if it makes things easier.

Reply:
Can't reproduce. Did you deploy before testing? (current deployed version is the current master).

And it would be a surprise if the owned could start the game before the others answered since the old game will only be destroyed when all leave (meaning sent `disconnect`, `leave`, or `rematch`)

Reply:
Can't reproduce now. I probably tested using wrong branch.